solr/core/src/java/org/apache/solr/handler/component/SortedHitQueueManager.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.handler.component;
+
+import org.apache.lucene.search.SortField;
+import org.apache.solr.search.AbstractReRankQuery;
+import org.apache.solr.search.RankQuery;
+import org.apache.solr.search.SortSpec;
+
+/**
+ * This class is used to manage the possible multiple SortedHitQueues that we need during mergeIds( ).
+ * Multiple queues are needed, if reRanking is used.
+ *
+ * If reRanking is disabled, only the queue is used.
+ * If reRanking is enabled, the top reRankDocsSize documents are added to the reRankQueue, all other documents are
+ * collected in the queue.
+ */
+public class SortedHitQueueManager {
+  private final ShardFieldSortedHitQueue queue;
+  private final ShardFieldSortedHitQueue reRankQueue;
+  private final int reRankDocsSize;
+
+  public SortedHitQueueManager(SortField[] sortFields, SortSpec ss, ResponseBuilder rb) {
+    final RankQuery rankQuery = rb.getRankQuery();
+
+    if(rb.getRankQuery() != null && rankQuery instanceof AbstractReRankQuery){
```

Comment:
> ... this pull request to fix for use cases with a score-free sort and future pull request(s) for use cases whose sort contains a score ...

I think something like this could work:

```suggestion
    boolean hasScoreSortField = false;
    for (SortField sortField : sortFields) {
      if (SortField.FIELD_SCORE.equals(sortField)) {
        hasScoreSortField = true;
        break;
      }
    }
    if(rb.getRankQuery() != null && rankQuery instanceof AbstractReRankQuery && hasScoreSortField == false){
```

What do you think? _(And no time pressure or rush at all to consider this idea, I just wanted post it here today to perhaps help us shift away for the "bit sad" conclusion of the "The current solution only works if ..." conversation above.)_

Reply:
Hi @cpoerschke ,
because your solution looked good to me, I added it in df6f30a . :)
I also added some comments about why we do that and linked to this PR.

Would you agree that this now finishes the scope of this PR?
I think this was the only ToDo left.

From my point of view the only thing left is to find a way of communicating what this PR solves and what is still broken. I do not know the right place for that, but maybe you have an idea?

Reply:
Hi @tomglk,

Thanks for updating the branch so that the existing behaviour is maintained when the sort includes a score!

In terms of communicating the scope of the fix, I can see a few complimentary possibilities:

* solr/CHANGES.txt entry for the JIRA issue: if this is very specific e.g. "custom sort without a score" rather than only "custom sort" wording then that could convey the scope and non-scope of the fix.

* Solr Ref Guide: https://solr.apache.org/guide/8_9/query-re-ranking.html appears to not yet mention distributed queries, so there's an opportunity there to document limitations or caveats e.g. https://solr.apache.org/guide/8_9/result-grouping.html#distributed-result-grouping-caveats does something similar.

* code javadocs: i've pushed an exploratory commit to the branch -- _feel free to revert or to adapt e.g. the (Single/Dual) SortedHitQueueManager name choice is probably too generic_ -- and in it the javadocs for the `DualSortedHitQueueManager.supports` method could provide a natural place to communicate what is and what is not supported and the usage of the method when deciding the DualSortedHitQueueManager vs. SingleSortedHitQueueManager instantiation helps communicate structually when existing behaviour is maintained since SingleSortedHitQueueManager is a simple wrapper around ShardFieldSortedHitQueue.

> ... I think this was the only ToDo left. ...

The test coverage in `solr/core` remains on the ToDo list I think, though it's easy to miss since it's rather hidden amongst all the discussion detail that led us to this point.

Reply:
Hi @cpoerschke ,
thank you for the different pointers for the documentation. :)

You are right, the test coverage still is a ToDo.

I like the approach of splitting the `SortedHitQueueManager`.
But I do not think that the QueryComponent should know when to create which SortedHitQueueManager. So I moved the `newSortedHitQueueManager()` into the manager itself and changed the interface to an abstract class. ( bbf887f )

What do you think about that change?

Reply:
> But I do not think that the QueryComponent should know when to create which SortedHitQueueManager. So I moved the `newSortedHitQueueManager()` into the manager itself and changed the interface to an abstract class. ([bbf887f](https://github.com/apache/solr/commit/bbf887f3bc4f24c1b5ff4ff213fb658f05c7d25f))
> 
> What do you think about that change?

Makes sense. And thanks for adding wonderful javadocs for the new classes too!